### PR TITLE
Module errors

### PR DESCRIPTION
**Pull request by aurianer**

### CODE REVIEW

File: hpx/custom_exception_info.hpp [hunk 1]
```diff
@@ -378,9 +100,52 @@ namespace hpx
...
+            }
+            catch (...)
+            {
+                return std::current_exception();
+            }
+
+            // need this return to silence a warning with icc
+            HPX_ASSERT(false);
+            return std::exception_ptr();
+        }
```
Comment (hkaiser) [inline hpx/custom_exception_info.hpp]: Is it necessary to _define_ this in a header file?

File: libs/errors/include/hpx/errors/exception_list.hpp [hunk 2]
```diff
@@ -35,7 +34,10 @@ namespace hpx
     class HPX_EXCEPTION_EXPORT exception_list : public hpx::exception
     {
     private:
-        typedef hpx::lcos::local::spinlock mutex_type;
+        // TODO: Does this need to be hpx::lcos::local::spinlock?
+        // typedef hpx::lcos::local::spinlock mutex_type;
+        // TODO: Add correct initialization of boost spinlock.
+        typedef boost::detail::spinlock mutex_type;
```
Comment (hkaiser) [inline libs/errors/include/hpx/errors/exception_list.hpp]: What's the rationale of this change?

File: hpx/custom_exception_info.hpp (hunk 1 above)
Comment (msimberg) [reply hpx/custom_exception_info.hpp]: Well, It *is* a templated function, but we might be able to get by with explicit instantiations. Not sure.

File: libs/errors/include/hpx/errors/exception_list.hpp (hunk 2 above)
Comment (msimberg) [reply libs/errors/include/hpx/errors/exception_list.hpp]: Not having to depend on `lcos` in this module (with the assumption that we don't actually need HPX thread yielding in the spinlock). But saying that I realize that `exception_list` might live more happily in the `algorithms` module in which case we should be able to keep it as `hpx::lcos::local::spinlock`. What do you think?

File: hpx/custom_exception_info.hpp (hunk 1 above)
Comment (hkaiser) [reply hpx/custom_exception_info.hpp]: We do explicitly instantiate this function (see: https://github.com/STEllAR-GROUP/hpx/blob/master/src/exception.cpp#L412-L461).

File: libs/errors/include/hpx/errors/exception_list.hpp (hunk 2 above)
Comment (hkaiser) [reply libs/errors/include/hpx/errors/exception_list.hpp]: Is it used only by the algorithms? I'm not sure. Also, assuming we don't need HPX yielding sounds like a bit of a stretch... However, I see your point...

Comment (msimberg) [reply libs/errors/include/hpx/errors/exception_list.hpp]: Hrm, almost only by algorithms (and executors)... two other files include it.

I know no HPX thread yielding does sound a bit like a stretch, but I don't see it as a performance problem since we're dealing with exceptions anyway in which case something has already gone wrong. That leaves the question whether this can deadlock. I can't prove this won't happen but I'm hoping our tests are extensive enough to tell us.

Having `lcos` as a dependency would make modularization very difficult so I'd rather try to make it work with the boost spinlock before resorting to that.

File: hpx/custom_exception_info.hpp (hunk 1 above)
Comment (msimberg) [reply hpx/custom_exception_info.hpp]: Sorry. Then this can definitely just be declared. Thanks!

File: libs/errors/include/hpx/errors/exception_list.hpp (hunk 2 above)
Comment (hkaiser) [reply libs/errors/include/hpx/errors/exception_list.hpp]: Ok, fair enough. BTW, there is also `hpx::util::spinlock` that relies on Boost's version but additionally does lock-tracking for HPX. So using this instead might actually be the better option.

Comment (msimberg) [reply libs/errors/include/hpx/errors/exception_list.hpp]: Agreed, but `hpx::util::spinlock` depends on exceptions making that problematic as well. This should one of, if not the, only module that would have to do this (depend directly on `boost::spinlock`). `hpx::util::spinlock` is coming up right after this module.

File: hpx/custom_exception_info.hpp (hunk 1 above)
Comment (msimberg) [reply hpx/custom_exception_info.hpp]: Looking a bit more carefully I see that it's `throw_exception` that's explicitly instantiated, not `construct_exception`. We would probably manage with the same explicit instantiations for `construct_exception` as well. Would you prefer that?

Comment (hkaiser) [reply hpx/custom_exception_info.hpp]: Yes, you are right. However, `construct_exception` is instantiated by `get_exception`, which is implicitly instantiated by `throw_exception` so there is no need to explicitly instantiate that as well.

Comment (msimberg) [reply hpx/custom_exception_info.hpp]: Sorry for the confusion (I'm slowly remembering the changes I made). That part has changed on this branch. `throw_exception` does not call `construct_exception` anymore. We can still instantiate `construct_exception` explicitly though.

Comment (hkaiser) [reply hpx/custom_exception_info.hpp]: Right, I should have looked more carefully. Let's go with explicitly instantiating the functions, then.

**Comment by msimberg** [comment]
The failure on CircleCI looks unrelated.
